README.md
# LSL-actiCHamp
LSL connector for the actiCHamp (plus) device from Brain Products.

To download, please click the 'Latest release' link to the right for the latest versions.

Please note that you may need to install the Microsoft C++ redistributable packages ([here](https://support.microsoft.com/en-us/help/2977003/the-latest-supported-visual-c-downloads)) in order to run the executables. Please note that for this app to function properly, you may need to download and install not only the package for Visual Studio 2015, 2017, and 2019, but also the Visual Studio 2012 (VC++ 11.0) redistributable package. Both packages can be found by following the link above.

# Getting Started
If you are new to using LSL, you might want to read the [Quick Start guide](https://labstreaminglayer.readthedocs.io/info/getting_started.html) on the LabStreamingLayer main repository. 

For measuring impedances, ensuring good signal quality, and troubleshooting basic amplifier functionality, it is always recommended to use [BrainVision Recorder](https://www.brainproducts.com/downloads.php?kid=2) and to consult the amplifier manual.

You can also find a useful (free) LSL viewer on the Brian Products website: [BrainVision LSL Viewer](https://www.brainproducts.com/downloads.php?kid=40&tab=3).

# Usage

Make sure that you have correctly installed the drivers for your amplifier, and that the amplifier is connected to the computer. Also make sure that the battery is fully charged and connected to the amplifier. It is always a good idea to be familiar with the amplifier manual before operation.

  * Start the actiCHamp Connector app. You should see a window like the following.
> ![actiCHamp.png](actiCHamp.png)

## Device Settings

### Device Serial Number

If you know the serial number of the amplifier you want to connect with, you may enter it manually in the 'Device Serial Number' field. Otherwise, you can scan for amplifiers using the 'Scan' button at the bottom of the GUI. After scanning, the'Available Devices' combo box will be populated with all available devices. The 'Device Serial Number' field will automatically be filled with the number of the first device found. Whichever device is referred to here will be the device that is connected with at the 'Link' action. 

### Number of EEG Channels

You can adjust the number of EEG channels that the device enables during Link action. Note that if the 'Overwrite on Channel Change' box is checked, changing these values will change the labels in the Channel Labels text box on the left side of the GUI. See the section on Channel Labels below for more information. The maximum number of available channels depends on the number of modules on the amplifier.

### Number of Aux Channels

The actiCHamp/actiCHamp Plus amplifiers have 8 available auxilary channels. Any number of these can be enabled (starting from channel 1 and moving up).

### Include Sample Ctr

Checking this box will add an additional channel to the EEG/AUX channel stream that counts the incoming samples coming from the device.

### Fast Data Access

The actiCHamp/actiCHamp Plus device is capable of very low-latency data delivery. However, in default mode (Fast Data Access unchecked) the driver will periodically probe the USB connection to ensure the cable hasn't been disconnected during data acquisition. If this safety check is disabled, it greatly reduces the amount of time between the moment a signal is acquired by the sensors and the time the digitally sample corresponding to the value at the sensor is available to the Connector app.

This is useful for low-latency applications such as closed-loop and BCI. It is recommended to use a relatively high sampling rate, e.g. 10 or 20kHz or else the length of the data buffer will be so large as to render this speedup negligible.

### Chunk Size

For most EEG experiments you can ignore the Chunk Size setting, but if you are developing a latency-critical real-time application (e.g., a P300 speller BCI), you can lower this setting to reduce the latency of your system. 

### Base Sampling Rate and Sub Sample Divisor

The actiCHamp/actiCHamp Plus has 3 native sampling frequencies: 100, 50, and 10kHz. The device driver can sub-sample in an extremely optimized manner by a factor of 1, 2, 5, 10, 20, 50, or 100. These values are the 'Base Sampling Rate' and 'Sub Sample Divisor' respectively. The Base Sampling Rate divided by the Sub Sample Divisor yields the the Nominal Sampling Rate.

### Nominal Sampling Rate

See above. Note that this is a read-only field. It will update when the Base Sampling Rate or Sub Sample Divisor is changed.

## LSL Trigger Output Style

These check boxes determine the way that the app handles device triggers in LSL.

Note that it is possible to enable neither, one of, or both trigger output styles simultaneously. They are not mutually exclusive.

### Unsampled String Markers

If 'Unsampled String Markers' is checked, the app will create a Marker stream on 'Link' that will forward the triggers received at the 1-bit trigger input on the hardware device. This stream will have the name of 'LiveAmpSN-xxxxxx-xxxx-DeviceTrigger' and its sampling rate is 0, i.e. it is a Marker stream 

If the STE box is connected, there will be two Marker streams created. One for the 1-bit amplifier trigger and one for the STE box trigger input. The STE Marker stream will have the name 'LiveAmpSN-xxxxxx-xxxx-STETriggerIn'.

If the STE box is connected and the STE Out Mode is set to 'Sync' there will be a third Marker outlet that corresponds to the hardware triggers sent from the trigger out of the STE box. The name of this Marker stream will be 'LiveAmpSN-xxxxxx-xxxx-STESync'. Note that this is not recommended for 'Unsampled String Markers' as it creates a large number of string markers (one per sample) which is not the intended use of LSL Marker outlets.

### EEG Channel

If this box is checked, extra channels will be added to the EEG/Bipolar/AUX/ACC stream corresponding to the requested trigger outputs. Rather than unsampled markers, these channels will output -1 if no trigger is available, else the value corresponding to the triggers in; and, in the case of the STE box in 'Sync' mode, the sync stream presented at the DSub output of the STE box. 

### Further Information

Please see the file [explanation_of_trigger_marker_types.pdf]( https://github.com/brain-products/LSL-LiveAmp/blob/master/explanation_of_trigger_marker_types.pdf) for more details.

 ## Link

 When pressed, the Link button will attempt to connect to the device specified in the 'Device Serial Number' field. When complete, the button text will change to 'Unlink' and all other GUI widgets will be disabled. An error message will pop up if no devices are available.

 When the Connector is linked to the actiCHamp, LSL streams will be created according to the settings set in the GUI. Please note that settings can be saved in a configuration file (see below) for ease of repetition.

## Scan For Devices

When 'Scan' is pressed, the app will search for available LiveAmps. When 1 or more actiCHamps is found, the Serial Numbers of these devices will populate the 'Available Devices' combo box. The first device found will automatically be placed in the 'Device Serial Number' field. When devices are available in the 'Available Devices' box, the user may choose from the available devices in the combo box by interacting with the combo box itself. The chosen device will be forwarded to the 'Device Number' field. If the selected device is unavailable at 'Link', the app will notify the user with an error message.

The LiveAmp device has a simulator mode which can be activated with the 'Use Simulator' checkbox. This will create 

## Configuration file

The configuration settings can be saved to a .cfg file (see File / Save Configuration) and subsequently loaded from such a file (via File / Load Configuration). Importantly, the program can be started with a command-line argument of the form "actiCHamp.exe -c myconfig.cfg", which allows to load the config automatically at start-up. The recommended procedure to use the app in experiments is to make a shortcut on the experimenter's desktop which points to a previously saved configuration customized to the study being recorded to minimize the chance of operator error.

## Active Shielding

There is no GUI setting to activate the Active Shielding option on the actiCHamp Connector. However, for those who know how to use this option (please read the amplifier manual before using this option!) you may do so by editing a configuration file in a text editor and including the following lines under the 'Settings' section:

`activeShield=true`

It is very important that you understand the consequences of activating this option as when used incorrectly, it can result in very unexpected results. Please consult the manual.

## Channel Labels


release_notes.txt
v1.15.2
-BUGFIX: change minimum EEG channel count from 1 to 0
-BUGFIX: prohibit attempts to have more than 32 channels of EEG at 100kHz and 64 at 50kHz 

v1.15.1
-BUGFIX: removed trigger mode setting error for old actiCHamp (not plus) 
-BUGFIX: config file loading properly

v1.15.0
-FEATURE: added mirror trigger output option
-FEATURE: report bitness in Help->About
-FEATURE: bumped lsl to version 1.15.2

v1.14.0
-FEATURE: added simulation mode support

v1.13.2
-BUGFIX: FDA was always on---now the toggle works properly

v1.13.1.1
-build number just for CI integration, no new features or bug fixes

actiCHamp LSL Connector v1.13.1
-BUGFIX: channel label typing bug
-BUGFIX: aux label update bug

actiCHamp LSL Connector v1.13.0
-added device scan option
-changed from AUX channels on/off to 0-8
-moved to INI config file format 
-disable all gui widgets on link

mainwindow.cpp
/*#define _CRTDBG_MAP_ALLOC  
#include <stdlib.h>  
#include <crtdbg.h> */ 

#include "mainwindow.h"
#include "ui_mainwindow.h"
#include "Amplifier_LIB.h"
#include <QtCore/QtCore>

#include <iostream>
#include <sstream>

#if _WIN64
#define BITDEPTH "64"
#else
#define BITDEPTH "32"
#endif

const int m_pnBaseSamplingRates[] = {10000,50000,100000};
const int m_pnSubSampleDivisors[] = { 1,2,5,10,20,50,100 };

void Transpose(const std::vector<std::vector<double> > &in, std::vector<std::vector<double> > &out);

#define APPNAME "actiCHamp LSL Connector " << BITDEPTH << "-bit"
#define LIBVERSIONSTREAM(version) version.Major << "." << version.Minor << "." << version.Build << "." << version.Revision
#define LSLVERSIONSTREAM(version) (version/100) << "." << (version%100)
#define APPVERSIONSTREAM(version) version.Major << "." << version.Minor << "." << version.Bugfix

MainWindow::MainWindow(QWidget *parent, const char* config_file): QMainWindow(parent),ui(new Ui::MainWindow)
{
	m_AppVersion.Major = 1;
	m_AppVersion.Minor = 15;
	m_AppVersion.Bugfix = 2;

	ui->setupUi(this);
	m_bUseActiveShield = false;
	m_bOverrideAutoUpdate = false;

	// make GUI connections
	QObject::connect(ui->actionQuit, SIGNAL(triggered()), this, SLOT(close()));
	QObject::connect(ui->linkButton, SIGNAL(clicked()), this, SLOT(Link()));
	QObject::connect(ui->actionLoad_Configuration, SIGNAL(triggered()), this, SLOT(LoadConfigDialog()));
	QObject::connect(ui->actionSave_Configuration, SIGNAL(triggered()), this, SLOT(SaveConfigDialog()));
	QObject::connect(ui->baseSamplingRate, SIGNAL(currentIndexChanged(int)), this, SLOT(SetSamplingRate()));
	QObject::connect(ui->subSampleDivisor, SIGNAL(currentIndexChanged(int)), this, SLOT(SetSamplingRate()));
	QObject::connect(ui->actionVersions, SIGNAL(triggered()), this, SLOT(VersionsDialog()));
	QObject::connect(ui->refreshDevices, SIGNAL(clicked()), this, SLOT(RefreshDevices()));
	QObject::connect(ui->eegChannelCount, SIGNAL(valueChanged(int)), this, SLOT(UpdateChannelLabelsEEG(int)));
	QObject::connect(ui->auxChannelCount, SIGNAL(valueChanged(int)), this, SLOT(UpdateChannelLabelsGUI(int)));
	QObject::connect(ui->availableDevices, SIGNAL(currentIndexChanged(int)), this, SLOT(ChooseDevice(int)));
	QObject::connect(ui->rbDefault, SIGNAL(clicked(bool)), this, SLOT(RadioButtonBehavior(bool)));
	QObject::connect(ui->rbMirror, SIGNAL(clicked(bool)), this, SLOT(RadioButtonBehavior(bool)));
	SetSamplingRate();
	QString cfgfilepath = find_config_file(config_file);
	LoadConfig(cfgfilepath);
}

QString MainWindow::find_config_file(const char* filename) {
	if (filename) {
		QString qfilename(filename);
		if (!QFileInfo::exists(qfilename))
			QMessageBox(QMessageBox::Warning, "Config file not found",
				QStringLiteral("The file '%1' doesn't exist").arg(qfilename), QMessageBox::Ok,
				this);
		else
			return qfilename;
	}
	QFileInfo exeInfo(QCoreApplication::applicationFilePath());
	QString defaultCfgFilename(exeInfo.completeBaseName() + ".cfg");
	qInfo() << defaultCfgFilename;
	QStringList cfgpaths;
	cfgpaths << QDir::currentPath()
		<< QStandardPaths::standardLocations(QStandardPaths::ConfigLocation) << exeInfo.path();
	for (auto path : cfgpaths) {
		QString cfgfilepath = path + QDir::separator() + defaultCfgFilename;
		if (QFileInfo::exists(cfgfilepath)) return cfgfilepath;
	}
	QMessageBox::warning(this, "No config file not found",
		QStringLiteral("No default config file could be found"), "Continue with default config");
	return "";
}

void MainWindow::VersionsDialog()
{
	t_VersionNumber libVersion;
	GetLibraryVersion(&libVersion);
	int32_t lslProtocolVersion = lsl::protocol_version();
	int32_t lslLibVersion = lsl::library_version();
	std::stringstream ss;
	ss << 
		"Amplifier_LIB: " << LIBVERSIONSTREAM(libVersion) << "\n" <<
		"lsl protocol: " << LSLVERSIONSTREAM(lslProtocolVersion) << "\n" <<
		"liblsl: " << LSLVERSIONSTREAM(lslLibVersion) << "\n" <<
		"App: " << APPVERSIONSTREAM(m_AppVersion) << ", " << BITDEPTH << "-bit";
	QMessageBox::information(this, "About", ss.str().c_str(), QMessageBox::Ok);
}

void MainWindow::SetSamplingRate()
{
	int nNum = m_pnBaseSamplingRates[ui->baseSamplingRate->currentIndex()];
	int nDenom = m_pnSubSampleDivisors[ui->subSampleDivisor->currentIndex()];
	std::string sSR = std::to_string(nNum / nDenom);
	ui->nominalSamplingRate->setText(sSR.c_str());
	m_nSamplingRate = nNum / nDenom;
	//setMinChunk();
}

void MainWindow::LoadConfigDialog() 
{
	QString filename = QFileDialog::getOpenFileName(this,"Load Configuration File","","Configuration Files (*.cfg)");
	if (!filename.isEmpty())
		LoadConfig(filename);
}

void MainWindow::SaveConfigDialog() 
{
	QString filename = QFileDialog::getSaveFileName(this, "Save Configuration File", "", "Configuration Files (*.cfg)");
	if (!filename.isEmpty())
		SaveConfig(filename);
}

void MainWindow::closeEvent(QCloseEvent *ev) 
{
	if (m_ptReadThread)
		ev->ignore();

}

int getBaseSamplingRateIndex(int nBaseSamplingRate)
{
	switch (nBaseSamplingRate)
	{
	case 10000:
		return 0;
	case 50000:
		return 1;
	case 100000:
		return 2;
	}
	return 0;
}

int getSubSampleDivisorIndex(int nSubSampleDivisor)
{
	switch (nSubSampleDivisor)
	{
	case 1:
		return 0;
	case 2:
		return 1;
	case 5:
		return 2;
	case 10:
		return 3;
	case 20:
		return 4;
	case 50:
		return 5;
	case 100:
		return 6;
	}
	return 0;
}

void MainWindow::RefreshDevices()
{
	ui->availableDevices->blockSignals(true);
	QStringList qsl;
	std::vector<std::pair<std::string, int>> ampData;
	this->setCursor(Qt::WaitCursor);
	this->setWindowTitle("Searching for Devices...");
	try
	{
		m_LibTalker.enumerate(ampData, ui->useSim->checkState() == Qt::Checked);
	}
	catch (std::exception & e)
	{
		QMessageBox::critical(this, "Error", (std::string("Could not locate actiCHamp device(s): ") += e.what()).c_str(), QMessageBox::Ok);
	}
	this->setCursor(Qt::ArrowCursor);
	if (!m_psAmpSns.empty())m_psAmpSns.clear();
	if (!m_pnUsableChannelsByDevice.empty()) m_pnUsableChannelsByDevice.clear();
	if (!ampData.empty()) {
		ui->availableDevices->clear();
		std::stringstream ss;
		int i = 0;
		for (std::vector<std::pair<std::string, int>>::iterator it = ampData.begin(); it != ampData.end(); ++it) {
			ss.clear();
			ss << it->first << " (" << it->second << ")";
			std::cout << it->first << std::endl;
			auto x = ss.str(); // oh, c++...
			qsl << QString(ss.str().c_str()); // oh, Qt...
			m_psAmpSns.push_back(it->first);
			m_pnUsableChannelsByDevice.push_back(it->second);
		}
		ui->availableDevices->addItems(qsl);
		ChooseDevice(0);
	}
	this->setWindowTitle("actiCHamp Connector");
	ui->availableDevices->blockSignals(true);
}

void MainWindow::UpdateChannelLabelsEEG(int n)
{

	if (m_bOverrideAutoUpdate)return;
	UpdateChannelLabels();
}

void MainWindow::UpdateChannelLabelsGUI(int n)
{

	if (m_bOverrideAutoUpdate)return;
	UpdateChannelLabels();
}


void MainWindow::UpdateChannelLabels()
{
	if (!ui->overwriteChannelLabels->isChecked())return;
	
	int nEeg = ui->eegChannelCount->value();
	int nAux = ui->auxChannelCount->value();

	std::string str;
	std::vector<std::string> psEEGChannelLabels;
	std::istringstream iss(ui->channelLabels->toPlainText().toStdString());
	int nEegChannelCount;
	while (std::getline(iss, str, '\n'))
		psEEGChannelLabels.push_back(str);
	nEegChannelCount = ui->eegChannelCount->value();

	for (auto it = psEEGChannelLabels.begin(); it != psEEGChannelLabels.end();)
	{
		if ((*(it)).find("AUX")!=std::string::npos)
			it = psEEGChannelLabels.erase(it);
		else
			++it;
	}

	while (psEEGChannelLabels.size() > nEegChannelCount)
		psEEGChannelLabels.pop_back();
	ui->channelLabels->clear();
	for (int i = 0; i < ui->eegChannelCount->value(); i++)
	{
		if (i < psEEGChannelLabels.size())
			str = psEEGChannelLabels[i];
		else
			str = std::to_string(i + 1);
		ui->channelLabels->appendPlainText(str.c_str());
	}
	for (int i = 0; i < nAux; i++)
	{
		str = "AUX_" + std::to_string(i + 1);
		ui->channelLabels->appendPlainText(str.c_str());
	}
}

void MainWindow::ChooseDevice(int which)
{

	if (!m_psAmpSns.empty())
		ui->serialNumber->setText(QString(m_psAmpSns[which].c_str()));
	ui->eegChannelCount->setMaximum(m_pnUsableChannelsByDevice[ui->availableDevices->currentIndex()]);
}

void MainWindow::RadioButtonBehavior(bool b)
{
	if (ui->rbMirror->isChecked())
	{
		m_TriggerOutputMode = TM_MIRROR;
	}
	if (ui->rbDefault->isChecked())
	{
		m_TriggerOutputMode = TM_DEFAULT;
	}
}

void MainWindow::LoadConfig(const QString &filename) 
{
	QSettings pt(filename, QSettings::IniFormat);

	ui->serialNumber->setText(pt.value("settings/serialNumber", "(enter manually or scan for devices)").toString());
	ui->eegChannelCount->setValue(pt.value("settings/eegChannelCount", 32).toInt());
	ui->auxChannelCount->setValue(pt.value("settings/auxChannelCount", 0).toInt());
	ui->chunkSize->setValue(pt.value("settings/chunkSize", 10).toInt());
	// TODO: make this a switch statement based on actual values
	int nBaseSamplingRate = pt.value("settings/baseSamplingRate", 10000).toInt();
	ui->baseSamplingRate->setCurrentIndex(getBaseSamplingRateIndex(nBaseSamplingRate));
	int nSubSampleDivisor = pt.value("settings/subSampleDivisor", 20).toInt();
	ui->nominalSamplingRate->setText((std::to_string(nBaseSamplingRate / nSubSampleDivisor)).c_str());
	ui->subSampleDivisor->setCurrentIndex(getSubSampleDivisorIndex(nSubSampleDivisor));
	//setMinChunk();
	ui->fastDataAccess->setCheckState(pt.value("settings/useFda", false).toBool() ? Qt::Checked : Qt::Unchecked);
	ui->useSampleCtr->setCheckState(pt.value("settings/useSampleCtr", false).toBool() ? Qt::Checked : Qt::Unchecked);
	m_bUseActiveShield = pt.value("settings/activeShield", false).toBool();
	ui->overwriteChannelLabels->setCheckState(pt.value("settings/overWrite", true).toBool() ? Qt::Checked : Qt::Unchecked);
	ui->sampledMarkersEEG->setCheckState(pt.value("settings/sampledMarkersEEG", false).toBool() ? Qt::Checked : Qt::Unchecked);
	ui->unsampledMarkers->setCheckState(pt.value("settings/unsampledMarkers", true).toBool() ? Qt::Checked : Qt::Unchecked);
	t_TriggerOutputMode triggerOutputMode = (t_TriggerOutputMode)pt.value("settings/triggerOutputMode", 0).toInt();
	if (triggerOutputMode == TM_MIRROR)ui->rbMirror->setChecked(true);
	else ui->rbDefault->setChecked(true);
	ui->channelLabels->clear();
	ui->channelLabels->setPlainText(pt.value("channels/labels").toStringList().join('\n'));
	UpdateChannelLabels();
	SetSamplingRate();

}

void MainWindow::SaveConfig(const QString& filename)
{
	QSettings pt(filename, QSettings::IniFormat);

	pt.beginGroup("settings");
	pt.setValue("serialNumber", ui->serialNumber->text());
	pt.setValue("eegChannelCount", ui->eegChannelCount->value());
	pt.setValue("auxChannelCount", ui->auxChannelCount->value());
	pt.setValue("useSampleCtr", ui->useSampleCtr->checkState() == Qt::Checked);
	pt.setValue("useFda", ui->fastDataAccess->checkState() == Qt::Checked);
	pt.setValue("chunkSize", ui->chunkSize->value());
	pt.setValue("baseSamplingRate", m_pnBaseSamplingRates[ui->baseSamplingRate->currentIndex()]);
	pt.setValue("subSampleDivisor", m_pnSubSampleDivisors[ui->subSampleDivisor->currentIndex()]);
	pt.setValue("unsampledMarkers", ui->unsampledMarkers->checkState() == Qt::Checked);
	pt.setValue("sampledMarkersEEG", ui->sampledMarkersEEG->checkState() == Qt::Checked);
	pt.setValue("overWrite", ui->overwriteChannelLabels->checkState() == Qt::Checked);
	pt.setValue("triggerOutputMode", m_TriggerOutputMode);
	pt.endGroup();
	pt.beginGroup("channels");
	pt.setValue("labels", ui->channelLabels->toPlainText().split('\n'));
	pt.endGroup();
}

void MainWindow::AmpSetup(t_AmpConfiguration ampConfiguration)
{
	m_LibTalker.setUseTriggers(ampConfiguration.m_bSampledMarkersEEG || ampConfiguration.m_bUnsampledMarkers);
	m_LibTalker.setExcludeTriggersFromOutput(!ui->sampledMarkersEEG->isChecked());
	m_LibTalker.setRequestedEEGChannelCnt(ampConfiguration.m_psEegChannelLabels.size());
	m_LibTalker.setRequestedAuxChannelCnt(ampConfiguration.m_psAuxChannelLabels.size());
	m_LibTalker.setBaseSamplingRate((float)ampConfiguration.m_nBaseSamplingRate);
	m_LibTalker.setSubSampleDivisor((float)ampConfiguration.m_nSubSampleDivisor);
	m_LibTalker.setUseFDA(ampConfiguration.m_bFDA);
	m_LibTalker.setUseSampleCtr(ampConfiguration.m_bUseSampleCtr);
	m_LibTalker.setUseActiveShield(false);
	m_LibTalker.Setup();
}

void MainWindow::ToggleGUIEnabled(bool bEnabled)
{
	if (bEnabled)
	{
		ui->channelLabels->setEnabled(true);
		ui->overwriteChannelLabels->setEnabled(true);
		ui->deviceGroup->setEnabled(true);
		ui->scanGroup->setEnabled(true);
		ui->channelGroup->setEnabled(true);
		ui->linkButton->setEnabled(true);
		ui->linkButton->setText("Link");
	}
	else
	{
		ui->channelLabels->setEnabled(false);
		ui->overwriteChannelLabels->setEnabled(false);
		ui->deviceGroup->setEnabled(false);
		ui->scanGroup->setEnabled(false);
		ui->channelGroup->setEnabled(false);
		ui->linkButton->setEnabled(true);
		ui->linkButton->setText("Unlink");
	}
}

void MainWindow::Link() 
{
	bool bSuccess = true;
	if (m_ptReadThread) 
	{
		try 
		{
			m_bStop = true;
			m_ptReadThread->join();
			m_ptReadThread.reset();
			m_LibTalker.setOutTriggerMode(TriggerOutputMode::TM_DEFAULT);
		} 
		catch(std::exception &e) 
		{
			QMessageBox::critical(this,"Error",(std::string("Could not stop the background processing: ")+=e.what()).c_str(),QMessageBox::Ok);
			return;
		}
		ToggleGUIEnabled(true);
		ui->linkButton->setText("Link");
	} 
	else 
	{
		try 
		{
			// get the UI parameters...
			t_AmpConfiguration ampConfiguration;
			ampConfiguration.m_sSerialNumber = ui->serialNumber->text().toStdString();
			ampConfiguration.m_bUseActiveShield = m_bUseActiveShield;
			ampConfiguration.m_bFDA = ui->fastDataAccess->checkState() == Qt::Checked;
			ampConfiguration.m_bUseSampleCtr = ui->useSampleCtr->checkState() == Qt::Checked;
			ampConfiguration.m_nChunkSize = ui->chunkSize->value();
			ampConfiguration.m_nBaseSamplingRate = ui->baseSamplingRate->currentText().toInt();
			ampConfiguration.m_nSubSampleDivisor = ui->subSampleDivisor->currentText().toInt();
			ampConfiguration.m_bUseSim = ui->useSim->checkState() == Qt::Checked;
			SetSamplingRate();
			std::vector<std::string> psChannelLabels;
			std::string str;
			std::istringstream iss(ui->channelLabels->toPlainText().toStdString());
			while (std::getline(iss, str, '\n'))
				psChannelLabels.push_back(str);
			std::vector<std::string> psEegChannelLabels;
			std::vector<std::string> psAuxChannelLabels;
			int i = 0;
			for (std::vector<std::string>::iterator it = psChannelLabels.begin();
				it < psChannelLabels.end();
				++it)
			{
				if (i < ui->eegChannelCount->value())
					psEegChannelLabels.push_back(*it);
				else
					psAuxChannelLabels.push_back(*it);
				i++;
			}
			ampConfiguration.m_psEegChannelLabels = psEegChannelLabels;
			ampConfiguration.m_psAuxChannelLabels = psAuxChannelLabels;
			ampConfiguration.m_bUnsampledMarkers = ui->unsampledMarkers->checkState() == Qt::Checked;
			ampConfiguration.m_bSampledMarkersEEG = ui->sampledMarkersEEG->checkState() == Qt::Checked;
			if (ampConfiguration.m_psEegChannelLabels.size() > 32 && (ampConfiguration.m_nBaseSamplingRate / ampConfiguration.m_nSubSampleDivisor == 100000))
			{
				QMessageBox::critical(this, "Error", "Can not initialize the actiCHamp. A Sampling Rate of 100kHz is limited to 32 channels of EEG.", QMessageBox::Ok);
				return;
			}
			if (ampConfiguration.m_psEegChannelLabels.size() > 64 && (ampConfiguration.m_nBaseSamplingRate / ampConfiguration.m_nSubSampleDivisor == 50000))
			{
				QMessageBox::critical(this, "Error", "Can not initialize the actiCHamp. A Sampling Rate of 50kHz is limited to 64 channels of EEG.", QMessageBox::Ok);
				return;
			}
			this->setWindowTitle(QString("Attempting to connect"));
			this->setCursor(Qt::WaitCursor);
			m_LibTalker.Connect(ampConfiguration.m_sSerialNumber, ampConfiguration.m_bUseSim);
			if(!m_LibTalker.setOutTriggerMode(m_TriggerOutputMode))
				ui->rbDefault->setChecked(true);
			AmpSetup(ampConfiguration);
			this->setCursor(Qt::ArrowCursor);
			this->setWindowTitle("actiCHamp Connector");
			
			// start reader thread
			m_bStop = false;
			m_ptReadThread.reset(new std::thread(&MainWindow::ReadThread,this, ampConfiguration));
		}
		catch(std::exception &e) 
		{

			std::string msg = e.what();
			QMessageBox::critical(this,"Error",("Could not initialize the actiCHamp: " + msg).c_str(),QMessageBox::Ok);	
			this->setCursor(Qt::ArrowCursor);
			ToggleGUIEnabled(true);
			return;
		}
		ToggleGUIEnabled(false);
		ui->linkButton->setText("Unlink");
	}
}



// background data reader thread
void MainWindow::ReadThread(t_AmpConfiguration ampConfiguration) 
{
	int res = SetPriorityClass(GetCurrentThread(), HIGH_PRIORITY_CLASS);
	//int nEnabledChannelCnt = m_nEEGChannelCount + ((m_bUseAUX) ? 8 : 0);
	int nChannelLabelCnt = ampConfiguration.m_psEegChannelLabels.size() + ampConfiguration.m_psAuxChannelLabels.size();
	int nExtraEEGChannelCnt = 0;
	//int nTriggerIdx = nChannelLabelCnt;
	//int nSampleCounterIdx = nTriggerIdx + 1;
	if (ampConfiguration.m_bSampledMarkersEEG) nExtraEEGChannelCnt+=1;
	if (ampConfiguration.m_bUseSampleCtr) nExtraEEGChannelCnt += 1;

	try
	{
		// for debugging:
		bool bFDA = m_LibTalker.CheckFDA();

		// setup LSL
		lsl::stream_outlet* poutMarkerOutlet = NULL;
		std::cout << std::string("actiCHamp" + m_LibTalker.getSerialNumber());
		std::cout << "EEG";
		std::cout << nChannelLabelCnt + nExtraEEGChannelCnt;
		std::cout << (double)m_nSamplingRate;
		std::cout << lsl::cf_float32;
		std::cout << m_LibTalker.getSerialNumber();
		lsl::stream_info infoData("actiCHamp-" + m_LibTalker.getSerialNumber(),
			"EEG",
			nChannelLabelCnt + nExtraEEGChannelCnt,
			(double)m_nSamplingRate,
			lsl::cf_float32,
			m_LibTalker.getSerialNumber());


		lsl::xml_element channels = infoData.desc().append_child("channels");
		for (int i = 0; i < ampConfiguration.m_psEegChannelLabels.size(); i++)
		{
			channels.append_child("channel")
				.append_child_value("label", ampConfiguration.m_psEegChannelLabels[i].c_str())
				.append_child_value("type", "EEG")
				.append_child_value("unit", "microvolts");
		}
		for (int i = 0; i < ampConfiguration.m_psAuxChannelLabels.size(); i++)
		{
			channels.append_child("channel")
				.append_child_value("label", ampConfiguration.m_psAuxChannelLabels[i].c_str())
				.append_child_value("type", "AUX")
				.append_child_value("unit", "microvolts");
		}
		if (ampConfiguration.m_bSampledMarkersEEG)
			channels.append_child("channel")
			.append_child_value("label", "Markers")
			.append_child_value("type", "Markers")
			.append_child_value("unit", "");

		if (ampConfiguration.m_bUseSampleCtr)
			channels.append_child("channel")
			.append_child_value("label", "SampleCounter")
			.append_child_value("type", "SampleCounter")
			.append_child_value("unit", "");

		infoData.desc().append_child("acquisition")
			.append_child_value("manufacturer", "Brain Products");

		t_VersionNumber libVersion;
		GetLibraryVersion(&libVersion);
		int32_t lslProtocolVersion = lsl::protocol_version();
		int32_t lslLibVersion = lsl::library_version();
		std::stringstream ssLib;
		ssLib << LIBVERSIONSTREAM(libVersion);
		std::stringstream ssProt;
		ssProt << LSLVERSIONSTREAM(lslProtocolVersion);
		std::stringstream ssLSL;
		ssLSL << LSLVERSIONSTREAM(lslLibVersion);
		std::stringstream ssApp;
		ssApp << APPVERSIONSTREAM(m_AppVersion);

		infoData.desc().append_child("versions")
			.append_child_value("Amplifier_LIB", ssLib.str())
			.append_child_value("lsl_protocol", ssProt.str())
			.append_child_value("liblsl", ssLSL.str())
			.append_child_value("App", ssApp.str());

		lsl::stream_outlet outData(infoData);

		lsl::stream_info infoMarkers("actiCHampMarkers-" + m_LibTalker.getSerialNumber(),
			"Markers",
			1,
			0,
			lsl::cf_string,
			m_LibTalker.getSerialNumber());

		if (ampConfiguration.m_bUnsampledMarkers)
		{
			infoMarkers.desc().append_child("channels")
				.append_child("channel")
				.append_child_value("label", "Markers")
				.append_child_value("type", "Markers")
				.append_child_value("unit", "");

			infoMarkers.desc().append_child("acquisition")
				.append_child_value("manufacturer", "Brain Products");

			infoMarkers.desc().append_child("versions")
				.append_child_value("Amplifier_LIB", ssLib.str())
				.append_child_value("lsl_protocol", ssProt.str())
				.append_child_value("liblsl", ssLSL.str())
				.append_child_value("App", ssApp.str());

			poutMarkerOutlet = new lsl::stream_outlet(infoMarkers);

		}
		std::vector<float> vfDataBufferMultiplexed;
		std::vector<int> vnTriggers;
		int nMarker;
		int nPreviousMarker = -1;
		int32_t nBufferSize = ampConfiguration.m_nChunkSize * m_LibTalker.getSampleSize();
		BYTE* buffer = NULL;
		buffer = new BYTE[nBufferSize];
		int64_t nSampleCnt;
		double dNow;
		int nSamplesPulled = 0;
		dNow = lsl::local_clock();
		vfDataBufferMultiplexed.resize(ampConfiguration.m_nChunkSize * (nChannelLabelCnt + nExtraEEGChannelCnt));
		vnTriggers.resize(ampConfiguration.m_nChunkSize);
		while (!m_bStop)
		{

			nSampleCnt = m_LibTalker.PullAmpData(buffer, nBufferSize, vfDataBufferMultiplexed, vnTriggers);
			if (nSampleCnt <= 0) {
				// CPU saver, this is ok even at higher sampling rates
				std::this_thread::sleep_for(std::chrono::milliseconds(1));
				continue;
			}
			nSamplesPulled += nSampleCnt;

			if (ampConfiguration.m_bUnsampledMarkers)
			{
				for (int s = 0; s < nSampleCnt; s++)
				{
					nMarker = (int)vnTriggers[s];
					if (nMarker != nPreviousMarker)
					{
						std::string sMarker = std::to_string(nMarker);
						poutMarkerOutlet->push_sample(&sMarker, dNow + (s + 1 - nSampleCnt) / m_nSamplingRate);
					}
					nPreviousMarker = nMarker;
				}
			}
			outData.push_chunk_multiplexed(vfDataBufferMultiplexed, dNow);
			vfDataBufferMultiplexed.clear();
			vnTriggers.clear();
			dNow = lsl::local_clock();
			nSamplesPulled = 0;
		}
		if (poutMarkerOutlet != NULL)
			delete(poutMarkerOutlet);
	}
	catch (std::exception & e)
	{
		std::string msg = e.what();
	}

}

MainWindow::~MainWindow() {
	delete ui;
}

